File: Java/src/Data_Structures/Arrays/MyArray.java
package Data_Structures.Arrays;

import java.util.*;

public class MyArray {
    int capacity;
    int length;
    Object[] data;


    public MyArray() {
        capacity = 1;
        length = 0;
        data = new Object[capacity];
    }

    // Display elements of the object
    void display()
    {
        System.out.println(Arrays.toString(this.data));
    }

    // Return length of the array
    int getLength()
    {
        return this.length;
    }

    // Return element at specified index
    Object getValue(int index) {
        if (index >= 0 && index < this.length)
        {
            return this.data[index];
        }
        throw new IndexOutOfBoundsException("Index out of range");
    }

    // Add element at the end of the array
    void push(Object value) {
        if (this.capacity == this.length)
        {
            this.capacity *= 2;
            this.data = Arrays.copyOf(this.data, capacity);
        }
        this.data[this.length] = value;
        this.length++;
    }

    // Remove and return last element of the array
    Object pop() {
        if (this.length > 0)
        {
            Object item = this.data[this.length - 1];
            this.data = Arrays.copyOf(this.data, this.length - 1);
            this.length--;
            this.capacity = this.length;
            return item;
        }

        throw new UnsupportedOperationException("Array is empty");
    }

    // Shift index of object to left
    void shift(Object[] objectDict, int size, int number) {
        for (int i = number; i < size - 1; i++) {
            objectDict[i] = objectDict[i + 1];
        }
    }

    // Delete element at specified index
    void delete(int index) {
        if (index >= 0 && index < this.length) {
            shift(this.data, this.length, index);
            this.data = Arrays.copyOf(this.data, this.length - 1);
            this.length--;
            this.capacity = this.length;
        } else {
            throw new IndexOutOfBoundsException("Index out of range");
        }
    }

    public static void main(String[] args) {
        MyArray arr = new MyArray();
        arr.push(45);
        arr.push(10);
        arr.push(50);
        arr.push(100);
        arr.push(27);
        System.out.println("size of array: " + arr.getLength());
        System.out.println("get elt at 1: " + arr.getValue(1));
        arr.display();
        System.out.println("removed last elt (27) and returned it: " + arr.pop());
        arr.display();
        arr.delete(2);
        arr.display();
    }
}

File: CSharp/Data_Structures/Arrays/MyArray.cs
namespace CSharp.Data_Structures.Arrays;

public class MyArray
{
    int capacity;
    int length;
    object[] data;

    public MyArray()
    {
        capacity = 1;
        length = 0;
        data = new object[capacity];
    }

    // Display elements of the object
    void Display()
    {
        Console.WriteLine("Array content:");
        for (int i = 0; i < length; i++)
        {
            Console.WriteLine(data[i]);
        }
    }

    // Return length of the array
    int GetLength()
    {
        return length;
    }

    // Return element at specified index
    object GetValue(int index)
    {
        if (index >= 0 && index < length)
        {
            return data[index];
        }
        throw new IndexOutOfRangeException("Index out of range");
    }

    // Add element at the end of the array
    void Push(object value)
    {
        if (capacity == length)
        {
            capacity *= 2;
            object[] newData = new object[capacity];
            Array.Copy(data, newData, length);
            data = newData;
        }
        data[length] = value;
        length++;
    }

    // Remove and return last element of the array
    object Pop()
    {
        if (length > 0)
        {
            object item = data[length - 1];
            object[] newData = new object[--length];
            Array.Copy(data, newData, length);
            data = newData;
            return item;
        }

        throw new InvalidOperationException("Array is empty");
    }

    // Shift index of object to the left
    void Shift(int index)
    {
        for (int i = index; i < length - 1; i++)
        {
            data[i] = data[i + 1];
        }
    }

    // Delete element at specified index
    void Delete(int index)
    {
        if (index >= 0 && index < length)
        {
            Shift(index);
            object[] newData = new object[--length];
            Array.Copy(data, newData, length);
            data = newData;
        }
        else
        {
            throw new IndexOutOfRangeException("Index out of range");
        }
    }

    public static void main(string[] args)
    {
        MyArray arr = new MyArray();
        arr.Push(45);
        arr.Push(10);
        arr.Push(50);
        arr.Push(100);
        arr.Push(27);
        Console.WriteLine("Size of array: " + arr.GetLength());
        Console.WriteLine("Get element at index 2: " + arr.GetValue(2));
        arr.Display();
        Console.WriteLine("Removed last element (27) and returned it: " + arr.Pop());
        arr.Display();
        arr.Delete(1); // Deletes the element at index 1
        arr.Display();
    }
}